test/js_test.go

```diff
@@ -2239,7 +2243,541 @@ func TestJetStream_UnsubscribeDeleteNoPermissions(t *testing.T) {
+		if recvd == 2 {
+			srvA.Shutdown()
+		}
+	}
+}
+
+func testJetStream_ClusterReconnectDurableQueueSubscriber(t *testing.T, subject string, srvs ...*jsServer) {
+	var (
+		srvA          = srvs[0]
+		srvB          = srvs[1]
+		srvC          = srvs[2]
+		totalMsgs     = 20
+		reconnected   = make(chan struct{})
+		reconnectDone bool
+	)
+	nc, err := nats.Connect(srvA.ClientURL(),
+		nats.ReconnectHandler(func(nc *nats.Conn) {
+			reconnected <- struct{}{}
+
+			// Bring back the server after the reconnect event.
+			if !reconnectDone {
+				reconnectDone = true
+				srvA.Restart()
+			}
+		}),
+	)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Drain to allow AckSync response to be received.
+	defer nc.Drain()
+
+	js, err := nc.JetStream()
+	if err != nil {
+		t.Error(err)
+	}
+
+	for i := 0; i < 10; i++ {
+		payload := fmt.Sprintf("i:%d", i)
+		js.Publish(subject, []byte(payload))
+	}
+
+	ctx, done := context.WithTimeout(context.Background(), 15*time.Second)
+	defer done()
+
+	msgs := make(chan *nats.Msg, totalMsgs)
+
+	// Create some queue subscribers.
+	srvAClientURL := srvA.ClientURL()
+	srvBClientURL := srvB.ClientURL()
+	srvCClientURL := srvC.ClientURL()
+	for i := 0; i < 5; i++ {
+		expected := totalMsgs
+		dname := "dur"
+		_, err = js.QueueSubscribe(subject, "wg", func(m *nats.Msg) {
+			msgs <- m
+
+			count := len(msgs)
+			switch {
+			case count == 2:
+				// Do not ack and wait for redelivery on reconnect.
+				srvA.Shutdown()
+				return
+			case count == 11:
+				// Do another Shutdown of the server we are connected with.
+				switch nc.ConnectedUrl() {
+				case srvAClientURL:
+					srvA.Shutdown()
+				case srvBClientURL:
+					srvB.Shutdown()
+				case srvCClientURL:
+					srvC.Shutdown()
+				default:
+				}
+				return
+			case count == expected:
+				done()
+			}
+
+			err := m.AckSync()
+			if err != nil {
+				// During the reconnection, both of these errors can occur.
+				if err == nats.ErrNoResponders || err == nats.ErrTimeout {
+					// Wait for reconnection event to occur to continue.
+					select {
+					case <-reconnected:
+						return
+					case <-time.After(1 * time.Second):
+						return
+					case <-ctx.Done():
+						return
+					}
+				}
+				t.Errorf("Unexpected error: %v", err)
+			}
+		}, nats.Durable(dname), nats.ManualAck())
+
+		if err != nil && err != nats.ErrTimeout {
+			t.Error(err)
+		}
+	}
+
+	var failedPubs int
+	for i := 10; i < totalMsgs; i++ {
+		var published bool
+		payload := fmt.Sprintf("i:%d", i)
+		timeout := time.Now().Add(10 * time.Second)
+
+	Retry:
+		for time.Now().Before(timeout) {
+			_, err = js.Publish(subject, []byte(payload))
+
+			// Skip temporary errors.
+			if err == nats.ErrNoStreamResponse || err == nats.ErrTimeout {
+				time.Sleep(100 * time.Millisecond)
+				continue Retry
+			} else if err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+			published = true
+			break Retry
+		}
+
+		if !published {
+			failedPubs++
```

Comment:
Some publishes can fail since we can lose quorum when restarting some of the servers, we account for this instead and  log a warning so changed so that test would fail only when not receiving the messages that were successfully published.